astropy/io/fits/hdu/image.py

```diff
@@ -214,6 +215,149 @@ def header(self, header):
+        dimension.  For example, an image of shape (100, 200) is defined to be
+        composed of 100 chunks.  As a result, slicing operations which limit
+        the data required along the first dimension will be more efficient.
+        For example, `.subset[10:12, :]` would trigger two chunks to be
+        downloaded, whereas `.subset[:, 10:12]` requires all chunks to
+        be downloaded.
+
+        This feature is intended to be used with files which are backed by a
+        (slow) remote data source. Specifically, this feature is intended
+        to enable subsets of FITS images hosted in AWS S3 cloud storage
+        to be accessed in an efficient way via the ``fsspec`` library.
+
+        This property does not offer data caching features.  It is assumed
+        that the underlying file object (``ImageHDU._file``) takes care of
+        caching redundant file reads.  This is usually the case when the
+        ``fsspec`` library is used to open the file objects.
+
+        You may wonder why the `.subset` property is introducing rather than
+        modifying the existing `ImageHDU.data` property.  This is the case
+        for two reasons:
+        * The `.data` property returns a `numpy.ndarray` object which translates
+          data slicing operations into the required file read calls. At face value
+          we may wish to build upon this. Unfortunately, these translations happens
+          deep in the NumPy C layer (in the case of the buffer protocol) or even in
+          the OS kernel (if memory mapping is used). There appears to be no obvious
+          way to call the pure Python code provided by ``fsspec`` in these layers.
+        * Although it would be possible to override `numpy.ndarray.__getitem__`
+          in pure Python, this is difficult because the instantiation
+          of an `ndarray` object appears to require the presence of a byte buffer
+          object that implements the Python Buffer Protocol (e.g., a memoryview object).
+          This is fundamentally a Python C API feature.  It is not clear
+          how an `fsspec`-backed buffer object can be implemented in pure Python.
+
+        Examples
+        --------
+        To extract a subset of a cloud-hosted FITS image:
+
+            >>> from astropy.io import fits
+            >>> uri = "s3://stpubdata/hst/public/j8pu/j8pu0y010/j8pu0y010_drc.fits"
+            >>> f = fits.open(uri)
+            >>> f[1].subset[10:12, 40:42]
+            array([[-0.02022914,  0.01017784],
+                   [-0.00605198,  0.01016048]], dtype=float32)
+
+        To configure the size of the blocks that are downloaded and cached by fsspec, you can
+        use the ``fsspec_kwargs`` parameter to pass on the relevant options:
+
+            >>> f = fits.open(uri, fsspec_kwargs={'default_cache_type': "block", 'default_block_size': 5_000_000})
+        """
+
+        class LazyLoadArray():
```

Comment:
I feel it may make sense to keep the reference to HDUs, maybe even image HDUs in the name of this class as it's not generic for any arrays

Reply:
Agreed!

Reply:
(I also intend to move this lazy array class somewhere else, TBD, as it's defined inline now!)

Reply:
I ended up being able to re-use the `ImageHDU.section` property rather than implementing a new `ImageHDU.subset` attribute.  As a result, this class is no longer needed and has been removed!